### PR TITLE
Generic code parameter in error correction trait

### PR DESCRIPTION
**Pull request by msoeken**
The `ErrorCorrection` trait is generalized such that it uses a code parameter instead of a code distance to compute the logical error rate, number of physical qubits, and cycle time. The code parameter's type is an associated type `Parameter` to the `ErrorCorrection` trait. It could be a numeric type, e.g., to represent a code distance, but it could also be a tuple type, to represent multiple parameters that influence the metrics above. A new trait function is added to traverse the possible space of values for the code parameter.

Further, this PR
- adds a test case to the system that checks consistency of the result output, in order to detect changes that break this format,
- removes serialization from the `estimate` folder and moves it to the `system` folder.

### CODE REVIEW

File: resource_estimator/src/system/data/report.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ impl Report {
...
-        entries.push(ReportEntry::new("physicalCountsFormatted/logicalErrorRate", "Logical qubit error rate", r#"Logical qubit error rate"#, &format!(r#"The logical qubit error rate is computed as ${} \cdot \left(\dfrac{{{}}}{{{}}}\right)^\frac{{{} + 1}}{{2}}$"#, job_params.qec_scheme().crossing_prefactor.expect("crossing prefactor should be set"), result.logical_qubit().physical_qubit().clifford_error_rate(), job_params.qec_scheme().error_correction_threshold.expect("error correction threshold should be set"), result.logical_qubit().code_distance())));
+        entries.push(ReportEntry::new("physicalCountsFormatted/logicalErrorRate", "Logical qubit error rate", r#"Logical qubit error rate"#, &format!(r#"The logical qubit error rate is computed as ${} \cdot \left(\dfrac{{{}}}{{{}}}\right)^\frac{{{} + 1}}{{2}}$"#, job_params.qec_scheme().crossing_prefactor.expect("crossing prefactor should be set"), result.logical_qubit().physical_qubit().clifford_error_rate(), job_params.qec_scheme().error_correction_threshold.expect("error correction threshold should be set"), result.logical_qubit().code_parameter())));
```
Comment (msoeken) [inline resource_estimator/src/system/data/report.rs:70]: Double check if code_distance needs to be renamed in markdown generation file.

File: resource_estimator/src/system/modeling/fault_tolerance.rs [hunk 2]
```diff
@@ -487,16 +489,16 @@ impl ErrorCorrection for Protocol {
     /// extraction time, is based on physical operation times specified in the
     /// qubit and usually some factor based on the choice of stabilizer
     /// extraction circuit.
-    fn logical_cycle_time(&self, qubit: &Self::Qubit, code_distance: u64) -> Result<u64, String> {
-        let mut context = Self::create_evaluation_context(Some(qubit), code_distance);
+    fn logical_cycle_time(&self, qubit: &Self::Qubit, code_distance: &u64) -> Result<u64, String> {
```
Comment (msoeken) [inline resource_estimator/src/system/modeling/fault_tolerance.rs:492]: Use associated type or assigned type (choose for consistency)

**Review by ivanbasov — APPROVED** [review]
reviewed on a sync

File: resource_estimator/src/system/data/report.rs (hunk 1 above)
Comment (msoeken) [reply resource_estimator/src/system/data/report.rs:70]: Double-checked. Generated output is based on JSON report structure, which is not changed by this PR.

File: resource_estimator/src/system/modeling/fault_tolerance.rs (hunk 2 above)
Comment (msoeken) [reply resource_estimator/src/system/modeling/fault_tolerance.rs:492]: Changed to use assigned type for `Self::Qubit`.

**Pull request merged**